File: apps/desktop/src/app/state/workspace-migration/security.ts
import type {
  DatastoreSecurityCheckSnapshot,
  DatastoreSecurityChecksPreferences,
  DatastoreSecurityFinding,
  DatastoreSecurityPostureCheckResult,
  DatastoreSecuritySeverity,
  DatastoreSecurityTarget,
  WorkspaceSnapshot,
} from '@datapadplusplus/shared-types'

export function normalizeDatastoreSecurityChecksPreferences(
  preferences: DatastoreSecurityChecksPreferences | undefined,
): DatastoreSecurityChecksPreferences {
  return {
    enabled: Boolean(preferences?.enabled),
    refreshIntervalDays: clampNumber(preferences?.refreshIntervalDays, 7, 1, 30),
    mutedFindingIds: normalizeStringList(preferences?.mutedFindingIds),
    lastRefreshAttemptAt:
      typeof preferences?.lastRefreshAttemptAt === 'string'
        ? preferences.lastRefreshAttemptAt
        : undefined,
    lastSuccessfulRefreshAt:
      typeof preferences?.lastSuccessfulRefreshAt === 'string'
        ? preferences.lastSuccessfulRefreshAt
        : undefined,
    nextManualRefreshAllowedAt:
      typeof preferences?.nextManualRefreshAllowedAt === 'string'
        ? preferences.nextManualRefreshAllowedAt
        : undefined,
  }
}

export function normalizeDatastoreSecurityCheckSnapshot(
  snapshot: WorkspaceSnapshot['datastoreSecurityChecks'] | undefined,
): DatastoreSecurityCheckSnapshot | undefined {
  if (!snapshot || typeof snapshot !== 'object') {
    return undefined
  }

  const status = ['idle', 'refreshing', 'ready', 'stale', 'error', 'unsupported'].includes(
    snapshot.status,
  )
    ? snapshot.status
    : 'idle'

  return {
    status,
    checkedAt: typeof snapshot.checkedAt === 'string' ? snapshot.checkedAt : undefined,
    expiresAt: typeof snapshot.expiresAt === 'string' ? snapshot.expiresAt : undefined,
    sourceMetadata: Array.isArray(snapshot.sourceMetadata)
      ? snapshot.sourceMetadata
          .filter((item) => item && typeof item === 'object')
          .map((item) => {
            const source: 'nvd' | 'cisa-kev' | 'version-catalog' =
              item.source === 'cisa-kev'
                ? 'cisa-kev'
                : item.source === 'version-catalog'
                  ? 'version-catalog'
                  : 'nvd'
            return {
              source,
              fetchedAt: typeof item.fetchedAt === 'string' ? item.fetchedAt : undefined,
              url: typeof item.url === 'string' ? item.url : '',
              recordCount: typeof item.recordCount === 'number' ? item.recordCount : undefined,
            }
          })
          .filter((item) => item.url)
      : [],
    targets: Array.isArray(snapshot.targets)
      ? snapshot.targets.map(normalizeDatastoreSecurityTarget)
      : [],
    findings: Array.isArray(snapshot.findings)
      ? snapshot.findings.map(normalizeDatastoreSecurityFinding)
      : [],
    postureChecks: Array.isArray(snapshot.postureChecks)
      ? snapshot.postureChecks.map(normalizeDatastoreSecurityPostureCheck)
      : [],
    warnings: normalizeStringList(snapshot.warnings),
    errors: normalizeStringList(snapshot.errors),
  }
}

function normalizeDatastoreSecurityTarget(
  target: Partial<DatastoreSecurityTarget>,
  index: number,
): DatastoreSecurityTarget {
  const status = [
    'pending',
    'checked',
    'notApplicable',
    'versionUnavailable',
    'mappingUnavailable',
    'error',
  ].includes(target.status ?? '')
    ? target.status
    : 'pending'
  const normalizedStatus = status as DatastoreSecurityTarget['status']

  return {
    id: typeof target.id === 'string' && target.id ? target.id : `security-target-${index + 1}`,
    connectionId: typeof target.connectionId === 'string' ? target.connectionId : '',
    environmentId: typeof target.environmentId === 'string' ? target.environmentId : '',
    connectionName: typeof target.connectionName === 'string' ? target.connectionName : 'Datastore',
    environmentName:
      typeof target.environmentName === 'string' ? target.environmentName : 'Environment',
    engine: typeof target.engine === 'string' ? target.engine : 'unknown',
    family: typeof target.family === 'string' ? target.family : 'unknown',
    status: normalizedStatus,
    detectedProduct:
      typeof target.detectedProduct === 'string' ? target.detectedProduct : undefined,
    detectedVersion:
      typeof target.detectedVersion === 'string' ? target.detectedVersion : undefined,
    knownLatestVersion:
      typeof target.knownLatestVersion === 'string' ? target.knownLatestVersion : undefined,
    recommendedVersion:
      typeof target.recommendedVersion === 'string' ? target.recommendedVersion : undefined,
    versionStatus: normalizeDatastoreVersionStatus(target.versionStatus),
    versionSource: normalizeDatastoreVersionSource(target.versionSource),
    versionSourceLabel:
      typeof target.versionSourceLabel === 'string' ? target.versionSourceLabel : undefined,
    versionSourceUrl:
      typeof target.versionSourceUrl === 'string' ? target.versionSourceUrl : undefined,
    versionSourceUpdatedAt:
      typeof target.versionSourceUpdatedAt === 'string' ? target.versionSourceUpdatedAt : undefined,
    cpeCandidates: Array.isArray(target.cpeCandidates)
      ? target.cpeCandidates
          .filter((candidate) => candidate && typeof candidate.cpeName === 'string')
          .map((candidate) => ({
            cpeName: candidate.cpeName,
            source: candidate.source === 'nvd' ? 'nvd' : 'curated',
            confidence:
              candidate.confidence === 'product' ||
              candidate.confidence === 'version-normalized'
                ? candidate.confidence
                : 'exact',
          }))
      : [],
    findingCount:
      typeof target.findingCount === 'number' && Number.isFinite(target.findingCount)
        ? Math.max(0, Math.floor(target.findingCount))
        : 0,
    highestSeverity: normalizeDatastoreSecuritySeverity(target.highestSeverity),
    lastCheckedAt: typeof target.lastCheckedAt === 'string' ? target.lastCheckedAt : undefined,
    message: typeof target.message === 'string' ? target.message : undefined,
    warnings: normalizeStringList(target.warnings),
  }
}

function normalizeDatastoreSecurityFinding(
  finding: Partial<DatastoreSecurityFinding>,
  index: number,
): DatastoreSecurityFinding {
  const cveId =
    typeof finding.cveId === 'string' && finding.cveId.trim()
      ? finding.cveId.trim()
      : `CVE-UNKNOWN-${index + 1}`
  return {
    id: typeof finding.id === 'string' && finding.id ? finding.id : cveId,
    targetIds: normalizeStringList(finding.targetIds),
    cveId,
    title: typeof finding.title === 'string' && finding.title ? finding.title : cveId,
    summary: typeof finding.summary === 'string' ? finding.summary : '',
    severity: normalizeDatastoreSecuritySeverity(finding.severity) ?? 'UNKNOWN',
    cvssScore:
      typeof finding.cvssScore === 'number' && Number.isFinite(finding.cvssScore)
        ? finding.cvssScore
        : undefined,
    cvssVector: typeof finding.cvssVector === 'string' ? finding.cvssVector : undefined,
    publishedAt: typeof finding.publishedAt === 'string' ? finding.publishedAt : undefined,
    modifiedAt: typeof finding.modifiedAt === 'string' ? finding.modifiedAt : undefined,
    affectedProduct:
      typeof finding.affectedProduct === 'string' ? finding.affectedProduct : 'Datastore',
    affectedVersion:
      typeof finding.affectedVersion === 'string' ? finding.affectedVersion : undefined,
    affectedVersionRange:
      typeof finding.affectedVersionRange === 'string'
        ? finding.affectedVersionRange
        : undefined,
    fixedVersionHint:
      typeof finding.fixedVersionHint === 'string' ? finding.fixedVersionHint : undefined,
    remediation:
      typeof finding.remediation === 'string' && finding.remediation.trim()
        ? finding.remediation
        : 'Review vendor guidance and apply a supported patched version.',
    references: Array.isArray(finding.references)
      ? finding.references
          .filter((reference) => reference && typeof reference.url === 'string')
          .map((reference) => ({
            label:
              typeof reference.label === 'string' && reference.label
                ? reference.label
                : reference.url,
            url: reference.url,
            source: typeof reference.source === 'string' ? reference.source : undefined,
          }))
      : [],
    cwes: normalizeStringList(finding.cwes),
    knownExploited: Boolean(finding.knownExploited),
    kev: finding.kev,
    sourceUrls: normalizeStringList(finding.sourceUrls),
  }
}

function normalizeDatastoreSecurityPostureCheck(
  check: Partial<DatastoreSecurityPostureCheckResult>,
  index: number,
): DatastoreSecurityPostureCheckResult {
  const ruleId =
    typeof check.ruleId === 'string' && check.ruleId.trim()
      ? check.ruleId.trim()
      : `posture.unknown.${index + 1}`
  return {
    id:
      typeof check.id === 'string' && check.id
        ? check.id
        : `posture-${ruleId.replace(/[^a-z0-9]+/gi, '-')}`,
    targetIds: normalizeStringList(check.targetIds),
    ruleId,
    category: normalizeDatastoreSecurityPostureCategory(check.category),
    status: normalizeDatastoreSecurityPostureStatus(check.status),
    severity: normalizeDatastoreSecuritySeverity(check.severity) ?? 'UNKNOWN',
    title: typeof check.title === 'string' && check.title ? check.title : ruleId,
    summary: typeof check.summary === 'string' ? check.summary : '',
    evidence: typeof check.evidence === 'string' ? check.evidence : undefined,
    remediation:
      typeof check.remediation === 'string' && check.remediation.trim()
        ? check.remediation
        : 'Review the datastore posture and apply least-privilege, authenticated, encrypted defaults where practical.',
    source: check.source === 'read-only-probe' ? 'read-only-probe' : 'profile',
    references: Array.isArray(check.references)
      ? check.references
          .filter((reference) => reference && typeof reference.url === 'string')
          .map((reference) => ({
            label:
              typeof reference.label === 'string' && reference.label
                ? reference.label
                : reference.url,
            url: reference.url,
            source: typeof reference.source === 'string' ? reference.source : undefined,
          }))
      : [],
  }
}

function normalizeDatastoreSecuritySeverity(
  severity: unknown,
): DatastoreSecuritySeverity | undefined {
  return severity === 'CRITICAL' ||
    severity === 'HIGH' ||
    severity === 'MEDIUM' ||
    severity === 'LOW' ||
    severity === 'NONE' ||
    severity === 'UNKNOWN'
    ? severity
    : undefined
}

function normalizeDatastoreVersionStatus(status: unknown) {
  return status === 'current' ||
    status === 'updateAvailable' ||
    status === 'unsupported' ||
    status === 'unknown'
    ? status
    : undefined
}

function normalizeDatastoreVersionSource(source: unknown) {
  return source === 'bundled-catalog' ||
    source === 'nvd-range' ||
    source === 'datastore-local'
    ? source
    : undefined
}

function normalizeDatastoreSecurityPostureStatus(status: unknown) {
  return status === 'pass' ||
    status === 'warn' ||
    status === 'fail' ||
    status === 'unknown' ||
    status === 'notApplicable'
    ? status
    : 'unknown'
}

function normalizeDatastoreSecurityPostureCategory(category: unknown) {
  return category === 'transport' ||
    category === 'auth' ||
    category === 'environment' ||
    category === 'secrets' ||
    category === 'privileges' ||
    category === 'durability' ||
    category === 'risky-settings' ||
    category === 'cloud' ||
    category === 'local-file'
    ? category
    : 'environment'
}

function clampNumber(value: unknown, fallback: number, min: number, max: number) {
  return typeof value === 'number' && Number.isFinite(value)
    ? Math.min(max, Math.max(min, Math.round(value)))
    : fallback
}

function normalizeStringList(values: unknown): string[] {
  return Array.isArray(values)
    ? [...new Set(values.filter((value): value is string => typeof value === 'string'))]
    : []
}

File: apps/desktop/src-tauri/src/domain/models/security.rs
use super::*;

#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[serde(default)]
pub struct DatastoreSecurityChecksPreferences {
    pub enabled: bool,
    pub refresh_interval_days: u32,
    pub muted_finding_ids: Vec<String>,
    pub last_refresh_attempt_at: Option<String>,
    pub last_successful_refresh_at: Option<String>,
    pub next_manual_refresh_allowed_at: Option<String>,
}

impl Default for DatastoreSecurityChecksPreferences {
    fn default() -> Self {
        Self {
            enabled: false,
            refresh_interval_days: 7,
            muted_finding_ids: Vec::new(),
            last_refresh_attempt_at: None,
            last_successful_refresh_at: None,
            next_manual_refresh_allowed_at: None,
        }
    }
}

#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DatastoreSecurityCpeCandidate {
    pub cpe_name: String,
    pub source: String,
    pub confidence: String,
}

#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DatastoreSecurityTarget {
    pub id: String,
    pub connection_id: String,
    pub environment_id: String,
    pub connection_name: String,
    pub environment_name: String,
    pub engine: String,
    pub family: String,
    pub status: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub detected_product: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub detected_version: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub known_latest_version: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub recommended_version: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version_status: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version_source: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version_source_label: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version_source_url: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub version_source_updated_at: Option<String>,
    #[serde(default)]
    pub cpe_candidates: Vec<DatastoreSecurityCpeCandidate>,
    pub finding_count: usize,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub highest_severity: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub last_checked_at: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
    #[serde(default)]
    pub warnings: Vec<String>,
}

#[derive(Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct DatastoreSecurityKevDetails {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub date_added: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub required_action: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub due_date: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub known_ransomware_campaign_use: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub notes: Option<String>,
}

#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DatastoreSecurityFindingReference {
    pub label: String,
    pub url: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source: Option<String>,
}

#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DatastoreSecurityFinding {
    pub id: String,
    #[serde(default)]
    pub target_ids: Vec<String>,
    pub cve_id: String,
    pub title: String,
    pub summary: String,
    pub severity: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cvss_score: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cvss_vector: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub published_at: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub modified_at: Option<String>,
    pub affected_product: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub affected_version: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub affected_version_range: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fixed_version_hint: Option<String>,
    pub remediation: String,
    #[serde(default)]
    pub references: Vec<DatastoreSecurityFindingReference>,
    #[serde(default)]
    pub cwes: Vec<String>,
    pub known_exploited: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub kev: Option<DatastoreSecurityKevDetails>,
    #[serde(default)]
    pub source_urls: Vec<String>,
}

#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DatastoreSecurityPostureCheckResult {
    pub id: String,
    #[serde(default)]
    pub target_ids: Vec<String>,
    pub rule_id: String,
    pub category: String,
    pub status: String,
    pub severity: String,
    pub title: String,
    pub summary: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub evidence: Option<String>,
    pub remediation: String,
    pub source: String,
    #[serde(default)]
    pub references: Vec<DatastoreSecurityFindingReference>,
}

#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DatastoreSecuritySourceMetadata {
    pub source: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fetched_at: Option<String>,
    pub url: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub record_count: Option<usize>,
}

#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DatastoreSecurityCheckSnapshot {
    pub status: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub checked_at: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<String>,
    #[serde(default)]
    pub source_metadata: Vec<DatastoreSecuritySourceMetadata>,
    #[serde(default)]
    pub targets: Vec<DatastoreSecurityTarget>,
    #[serde(default)]
    pub findings: Vec<DatastoreSecurityFinding>,
    #[serde(default)]
    pub posture_checks: Vec<DatastoreSecurityPostureCheckResult>,
    #[serde(default)]
    pub warnings: Vec<String>,
    #[serde(default)]
    pub errors: Vec<String>,
}

#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DatastoreSecurityChecksSettingsRequest {
    pub enabled: bool,
    pub refresh_interval_days: Option<u32>,
    pub muted_finding_ids: Option<Vec<String>>,
}

#[derive(Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct DatastoreSecurityChecksRefreshRequest {
    #[serde(default)]
    pub manual: bool,
}

#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DatastoreSecurityChecksStatus {
    pub supported: bool,
    pub enabled: bool,
    pub message: String,
    pub can_refresh: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub refresh_blocked_reason: Option<String>,
    pub preferences: DatastoreSecurityChecksPreferences,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub snapshot: Option<DatastoreSecurityCheckSnapshot>,
}
